budgetportal/tests/test_prov_infra_projects.py

```diff
@@ -1,247 +1,959 @@
+        )
+        self.project = ProvInfraProject.objects.create(IRM_project_id=123456)
+        self.project_snapshot = ProvInfraProjectSnapshot.objects.create(
+            irm_snapshot=self.irm_snapshot,
+            project=self.project,
+            name="BLUE JUNIOR SECONDARY SCHOOL",
+            # Administration
+            department="Education",
+            budget_programme="Programme 2 - Public Ordinary School Education",
+            project_number="W/50042423/WS",
+            status="Construction",
+            # Location
+            province="KwaZulu-Natal",
+            local_municipality="Dr Nkosazana Dlamini Zuma",
+            district_municipality="Harry Gwala",
+            # Implementation
+            program_implementing_agent="DOPW",
+            principle_agent="PRINCIPLE AGENT",
+            main_contractor="MAIN CONTRACTOR",
+            other_parties="OTHERS",
+            # Funding
+            primary_funding_source="Education Infrastructure Grant",
+            nature_of_investment="Upgrading and Additions",
+            funding_status="Tabled",
+            # Budget
+            total_project_cost=680000,
+            total_construction_costs=562000,
+            total_professional_fees=118000,
+            # Cost to date
+            expenditure_from_previous_years_total=556479,
+            expenditure_from_previous_years_professional_fees=118000,
+            expenditure_from_previous_years_construction_costs=0,
+            variation_orders=0,
+            # Original Budget FY
+            main_appropriation_total=337000,
+            main_appropriation_construction_costs=276000,
+            main_appropriation_professional_fees=61000,
+            # Adjustment Budget FY
+            adjustment_appropriation_total=1,
+            adjustment_appropriation_construction_costs=2,
+            adjustment_appropriation_professional_fees=3,
+            # Overall timeline
+            start_date=date(2016, 6, 13),
+            estimated_completion_date=date(year=2021, month=6, day=30),
+            # Construction timeline
+            estimated_construction_start_date=date(2017, 2, 1),
+            estimated_construction_end_date=date(2020, 12, 31),
+            contracted_construction_end_date=date(2021, 1, 1)
+        )
 
+        ProvInfraProjectIndex().reindex()
+
+    def tearDown(self):
+        ProvInfraProjectIndex().clear()
+
+    def test_project_detail_page_fields(self):
         selenium = self.selenium
+        url = self.project.get_absolute_url()
+        selenium.get("%s%s" % (self.live_server_url, url))
+        title = selenium.find_element_by_css_selector(".page-heading").text
+        self.assertEqual(title, u"BLUE JUNIOR SECONDARY SCHOOL")
+
+        funding = selenium.find_element_by_xpath(
+            "//html/body/div[4]/div/div[2]/div[1]/div"
+        )
+        total_cost = funding.find_element_by_xpath(".//div[2]/div[2]").text
```

Comment:
Please use the selectors listed here for all these fields, rather than using the markup - the selectors are required to stay the same between webflow and django, so that they html can change without breaking anything.

e.g. `selenium.find_element_by_css_selector(...)`
Using XPath like this will break frequently because the HTML will change

Reply:
See see the comments about `wait.until` below - that will help you write cleaner, clearer and more robust tests here

Reply:
all find_element_by_xpath methods changed with find_element_by_css_selector